### PR TITLE
mcomix3: init at 20201123

### LINKED ISSUE #75028: Package mcomix3
**Describe the bug**
There is a fork of mcomix, mcomix3, that's is currently in use by debian because it's ported to gtk3 and python3.

https://github.com/multiSnow/mcomix3/

A big issue though is it lacks a real build system to install
* https://github.com/multiSnow/mcomix3/issues/93

If anyone is up to maintaining this I could contribute https://github.com/multiSnow/mcomix3/issues/96.

cc @fuuzetsu @AndersonTorres.

### PR DESCRIPTION
**Pull request by con-f-use**
###### Motivation for this change
The [original mcomix](https://sourceforge.net/projects/mcomix/) used to be in nixpkgs, but became unmaintained and is not Python 3 compatible. It was subsequently [removed](https://github.com/NixOS/nixpkgs/commit/28d42831217fb8661c3b7a5756998326f0605360). Debian, Archlinux, Fedora and others have switched to it's actively maintained Python 3 fork, [mcomix3](https://github.com/multiSnow/mcomix3). This PR packages mcomix3 for nixpkgs.

It is @con-f-use first contribution to nixpkgs, so apologies for any errors or breach of protocol. 

Closes #75028.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/graphics/mcomix3/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/graphics/mcomix3/default.nix:31]:
```suggestion
  propagatedBuildInputs = (with python3Packages; [ pillow pygobject3 pycairo ])
    ++ [ gobject-introspection gtk3 gdk-pixbuf unrar p7zip lhasa mupdf ];
```

File: pkgs/applications/graphics/mcomix3/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/graphics/mcomix3/default.nix:74]:
```suggestion
  # The tests seem to be broken upstream
  doCheck = false;

  checkInputs = [ python3Packages.pytest ];

  checkPhase = ''
    HOME=$TMPDIR pytest
  '';

  pythonImportsCheck = [ "mcomix" ];
```

File: pkgs/applications/graphics/mcomix3/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/graphics/mcomix3/default.nix:81]: IIRC because of the unrar dependency people need to enable unfree packages.

File: pkgs/top-level/aliases.nix
Comment (SuperSandro2000) [inline pkgs/top-level/aliases.nix:310]:
```suggestion
  mcomix = throw "mcomix has been removed from nixpkgs, as it's unmaintained; try mcomix3 a Python 3 fork"; # added 2019-12-10, modified 2020-11-25
```

File: pkgs/applications/graphics/mcomix3/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/graphics/mcomix3/default.nix:10]: If they are runtime dependencies we could wrap them. Not sure if this is nicer though.

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 104908` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mcomix3</li>
  </ul>
</details>

File: pkgs/applications/graphics/mcomix3/default.nix
Comment (ryantm) [reply pkgs/applications/graphics/mcomix3/default.nix:81]: I don't think we need to make something unfree because it dependency is.

File: pkgs/applications/graphics/mcomix3/default.nix
Comment (con-f-use) [reply pkgs/applications/graphics/mcomix3/default.nix:10]: They are runtime.

File: pkgs/applications/graphics/mcomix3/default.nix
Comment (SuperSandro2000) [reply pkgs/applications/graphics/mcomix3/default.nix:81]: I didn#t want to suggest that but people can't build it when they don't enable unfree packages, right?

File: pkgs/applications/graphics/mcomix3/default.nix
Comment (SuperSandro2000) [reply pkgs/applications/graphics/mcomix3/default.nix:10]: Something like this:

```
wrapProgram $out/bin/somafm --prefix PATH ":" "${stdenv.lib.makeBinPath [ curl jq mpv ]}";
```

File: pkgs/applications/graphics/mcomix3/default.nix
Comment (jonringer) [reply pkgs/applications/graphics/mcomix3/default.nix:81]: this should work transitively, `nix-build` or related tool will complain that you need to allow unfree

File: pkgs/applications/graphics/mcomix3/default.nix
Comment (jtojnar) [inline pkgs/applications/graphics/mcomix3/default.nix:30]: The introspection libraries should not need to be propagated.

File: pkgs/applications/graphics/mcomix3/default.nix
Comment (jtojnar) [inline pkgs/applications/graphics/mcomix3/default.nix:31]: And propagating these will not really make these accessible.

File: pkgs/applications/graphics/mcomix3/default.nix
Comment (jtojnar) [inline pkgs/applications/graphics/mcomix3/default.nix:45]: Looks like https://github.com/multiSnow/mcomix3/issues/93 still is not resolved :-/

Comment (con-f-use) [reply pkgs/applications/graphics/mcomix3/default.nix:48]: Yes, and sadly, it likely never will. So I figured this is better than not having mcomix, wouldn't you agree?

File: pkgs/applications/graphics/mcomix3/default.nix
Comment (con-f-use) [reply pkgs/applications/graphics/mcomix3/default.nix]: > The introspection libraries should not need to be propagated.
> And propagating these will not really make these accessible.

Okay, so how do I do it correctly?

```
  buildInputs = [ gobject-introspection p7zip lhasa mupdf ]
    ++ lib.optional (unrarSupport) unrar;
  nativeBuildInputs = [ wrapGAppsHook ];
  propagatedBuildInputs = (with python3Packages; [ pillow pygobject3 pycairo ])
    ++ [ gtk3 gdk-pixbuf ];
```

Is this more correct? I tested it and it seems to work.

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 104908` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mcomix3</li>
  </ul>
</details>

File: pkgs/applications/graphics/mcomix3/default.nix
Comment (AndersonTorres) [inline pkgs/applications/graphics/mcomix3/default.nix:73]: Why the name is `comicthump` and not `comicthumb`?

File: pkgs/applications/graphics/mcomix3/default.nix
Comment (AndersonTorres) [reply pkgs/applications/graphics/mcomix3/default.nix:82]: Well, some things:
- unrar is a binary program, not a dinamically linked library;
- comix remains basically the same with or without unrar. It doesn't produce a new binary program just because it calls unrar if present in the "path".

That being said, the license is still GPL despite it calls unrar or not.

File: pkgs/applications/graphics/mcomix3/default.nix
Comment (AndersonTorres) [inline pkgs/applications/graphics/mcomix3/default.nix:79]:
```suggestion
    longDescription = "User-friendly, customizable image viewer, specifically designed to handle comic books and manga supporting a variety of container formats (including CBR, CBZ, CB7, CBT, LHA and PDF)";
```

```suggestion
    longDescription = ''
--------------------------------------------------------------------------------     
      User-friendly, customizable image viewer, specifically designed to handle comic books 
      and manga supporting a variety of container formats (including CBR, CBZ, CB7, CBT, LHA and PDF)".
```

Use the two-quote notation **and** limit the lines to 80 columns:

Comment (con-f-use) [reply pkgs/applications/graphics/mcomix3/default.nix:79]:
```suggestion
    longDescription = ''
      User-friendly, customizable image viewer, specifically designed to handle
      comic books and manga supporting a variety of container formats
      (including CBR, CBZ, CB7, CBT, LHA and PDF)
    '';
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
Now squash all those small commits to a single commit.

Or, more accurately, two: 
- the first including yourself in the maintainers list
- the other including all else, namely, the package expression.

**Comment by con-f-use** [comment]
Okay, little help with that please, because if I do `git branch --set-upstream-to=origin/master; git rebase -i`, there is 2000+ commits in that interactive rebase and hundreds of merge conflicts to resolve.

File: pkgs/applications/graphics/mcomix3/default.nix
Comment (ryantm) [reply pkgs/applications/graphics/mcomix3/default.nix:86]: @SuperSandro2000 Right, they won't be able to build this.
@con-f-use If unrar is critical to how it works, I'd say leave it how it is. If not, then your `unrarSupport` change sounds good.

**Comment by AndersonTorres** [comment]
Usually I just
- fetch upstream;
- rebase upstream/master

After this, the branch is on par with upstream Nixpkgs.

**Comment by jtojnar** [comment]
But just adding it to buildInputs will not be sufficient. That will 
only cause it to be on PATH during the build. If you want it to be 
available at runtime as well, you will need to use wrappers.

**Comment by con-f-use** [comment]
Okay, I'll look into wrapping. Is there a good way, so it doesn't interfere with the wrapping from pythonApplication and `wrapGAppsHook`?

**Comment by AndersonTorres** [comment]
I think you will need to wrap the things manually. First, `dontWrapGApps = true;`

https://github.com/NixOS/nixpkgs/blob/ad8238c483133e617fe6dfc8855449c029a3775a/pkgs/applications/misc/genxword/default.nix#L39

**Comment by con-f-use** [comment]
Well, I don't seem to **need** to, because multiple layers of wrapping are working, the environment variables are all there, once I get to the actual mcomix3 executable, but it **would** be nicer.

**Comment by jtojnar** [comment]
You can just add stuff to `makeWrapperArgs`, which will be picked up by the Python wrapper. If you combine it with https://github.com/NixOS/nixpkgs/pull/104908#issuecomment-734236008, the python wrapper will also take the data from `wrapGAppsHook` so you will end up with just a single wrapper.

But yeah, it is a trade off between having nested wrappers and slightly more complicated package expressions.

File: pkgs/applications/graphics/mcomix3/default.nix
Comment (jtojnar) [inline pkgs/applications/graphics/mcomix3/default.nix:34]: These two still should not be propagated, only Python dependencies need to be and [soon not even those](https://github.com/NixOS/nixpkgs/pull/102613).

File: pkgs/applications/graphics/mcomix3/default.nix
Comment (jtojnar) [inline pkgs/applications/graphics/mcomix3/default.nix:20]: According to https://nixos.org/nixpkgs/manual/#sec-package-naming, the `version` should be `unstable-yyyy-mm-dd`, when using git snapshots.

Comment (con-f-use) [reply pkgs/applications/graphics/mcomix3/default.nix:20]: Okay, everybody is telling me something different about what the version should be. So far, I had `1.2.1+git20201123`, `v1.2.1+git<commit>`, `20201123` your `unstable-2020-11-23`. I'll take yours, because you provided a manual link.

File: pkgs/applications/graphics/mcomix3/default.nix
Comment (con-f-use) [reply pkgs/applications/graphics/mcomix3/default.nix:34]: Weird, I could've sworn that I had them in `buildInputs` before and things were not working then. That's why I moved them at some point. But now that I moved them again, it's fine. I'm very happy, when this thing is eventually merged, I'm starting to go crazy.

**Comment by con-f-use** [comment]
```nix
  dontWrapGApps = true;
  preFixup = ''
    makeWrapperArgs+=(
      "''${gappsWrapperArgs[@]}"
      "--prefix" "PATH" ":" "${lib.makeBinPath ([ p7zip lhasa mupdf ] ++ lib.optional (unrarSupport) unrar)}" 
    )
  '';
```
The `--prefix` part doesn't seem to work with the python wrapper. And when I set the attribute directly in the nix expression like so:

```
   makeWrapperArgs = [ "--prefix PATH : ${lib.makeBinPath ([ p7zip lhasa mupdf ] ++ lib.optional (unrarSupport) unrar)}" ];
```

Then the arguments from `wrapGAppsHook` don't get passed anymore. Any pointers here?

**Comment by jtojnar** [comment]
You cannot set `makeWrapperArgs` in Nix directly, it will get mangled when serialized for bash.

What do you mean the prefix does not work? Is it not appearing in the created wrapper?

**Comment by con-f-use** [comment]
Nevermind, I think, I'm finally there. Thanks for your insight!

File: pkgs/applications/graphics/mcomix3/default.nix
Comment (jtojnar) [inline pkgs/applications/graphics/mcomix3/default.nix:5]: Use either `python3Packages.python` with `python3Packages.foomodule`, or `python3` with `python3.pkgs.foomodule`. Mixing the two styles can lead to inconsistencies when one is overridden.

File: pkgs/applications/graphics/mcomix3/default.nix
Comment (jtojnar) [inline pkgs/applications/graphics/mcomix3/default.nix:36]: It would be nice to add a comment like the other packages do.

**Comment by con-f-use** [comment]
Okay, I think I have followed all the suggestions. @jtojnar , please tell me if you now thing it's ready to be merged, then I'll squash the commits again, and hopefully be done.

**Comment by jtojnar** [comment]
Yes it looks okay to me now. Thanks.

**Comment by con-f-use** [comment]
Okay, commits are squashed, what need to happen now to get this thing merged?

**Comment by jtojnar** [comment]
I can merge this once CI finishes. Thanks.

**Comment by con-f-use** [comment]
Thank **you**, I learned a lot. CI seems finished...

**Comment by jtojnar** [comment]
Thanks.

**Pull request merged**
